File: projects/inflection/modules/dev/dev_lua/tools.lua
local export = {}

local u = require('Module:utils')

local function trim_stress(str)
	-- remove space after stress (we put it in the code just to make visually readable sources)
	local result, count = mw.ustring.gsub(str, '́ ', '́')
	return result
end

export.stash = {}
level = 0

function export.clear_stash()
	export.stash = {}
end

function export.add_stash(name, value)
	export.stash[name] = value
end

function export.apply_stash(str)
	for name, value in pairs(export.stash) do
		str = mw.ustring.gsub(str, u.escape(name), value)
	end
	return str
end

function export.replaced(str, pattern, replace_to)
	pattern = export.apply_stash(pattern)
	return mw.ustring.gsub(str, trim_stress(pattern), trim_stress(replace_to))
end

function export.replace(dict, key, pattern, replace_to)
	if key == 'all-sg' then
		local keys = {'gen-sg', 'dat-sg', 'prp-sg' }  -- without 'nom-sg', 'acc-sg' and 'ins-sg'
		for i, key in pairs(keys) do
			dict[key] = export.replaced(dict[key], pattern, replace_to)
		end
	elseif key == 'all-pl' then
		local keys = {'nom-pl', 'gen-pl', 'dat-pl', 'ins-pl', 'prp-pl' }  -- without 'acc-pl'
		for i, key in pairs(keys) do
			dict[key] = export.replaced(dict[key], pattern, replace_to)
		end
	else
		dict[key] = export.replaced(dict[key], pattern, replace_to)
	end
end

function export.extract(str, pattern)
	pattern = export.apply_stash(pattern)
	return mw.ustring.match(str, trim_stress(pattern))
end

function export.check(str, values, checker)
	if type(values) == 'string' then
		values = {values}
	end
	for i, value in pairs(values) do
		value = trim_stress(export.apply_stash(value))
		local ok =
			   checker == 'equals'           and value == str
			or checker == 'startswith'       and mw.ustring.match(str, "^" .. value) ~= nil
			or checker == 'endswith'         and mw.ustring.match(str, value .. "$") ~= nil
			or checker == 'penultimate'      and mw.ustring.match(str, value .. ".$") ~= nil
			or checker == 'contains'         and mw.ustring.match(str, value) ~= nil
			or checker == 'contains_once'    and table.getn(mw.text.split(str, value)) == 2
			or checker == 'contains_several' and table.getn(mw.text.split(str, value)) > 2
		if ok then
			return true
		end
	end
	return false
end

function export.equals(str, values)
	return export.check(str, values, 'equals')
end

function export.In(str, values)
	return export.equals(str, values)
end

function export.startswith(str, values)
	return export.check(str, values, 'startswith')
end

function export.endswith(str, values)
	return export.check(str, values, 'endswith')
end

function export.penultimate(str, values)
	return export.check(str, values, 'penultimate')
end

function export.contains(str, values)
	return export.check(str, values, 'contains')
end

function export.contains_once(str, values)
	return export.check(str, values, 'contains_once')
end

function export.contains_several(str, values)
	return export.check(str, values, 'contains_several')
end

-- TODO
--function export.log(args, ...)
--	if nil then
--		'ERROR'
--	end
--end

function export.log(value)
    local prefix = string.rep(' ', level)
    mw.log(prefix .. value)
end

function export.log_info(info)
    export.log('# ' .. info)
end

function export.call(module, name)
    export.log('@ ' .. module .. '.' .. name .. '()')
end

function export.starts(module, name)
    export.log('↘ @ ' .. module .. '.' .. name .. '():')
    level = level + 4
end

function export.ends(module, name)
    level = level - 4
    -- log('↙ . ' .. module .. '.' .. name .. '()')
    export.log('↙ . ')
end

function export.returns(module, name, result)
    level = level - 4
    -- log('↙ . ' .. module .. '.' .. name .. '()')
    export.log('↙ . ')
    return result
end

function export.log_value(value, name)
    export.log('= ' .. name .. ': "' .. tostring(value) .. '"')
end

function export.log_table(t, name)
	export.log('- ' .. name .. ':')
	for key, value in pairs(t) do
		export.log('  ["' .. tostring(key) .. '"] = "' .. tostring(value) .. '"')
	end
end

function export.has_key(value)
	-- just return value itself (need for python conversion scripts)
	return value
end

function export.has_value(value)
	-- just check if variable `value` has non-empty value
	return value and value ~= ''
end

function export.set(value)  -- FIXME: Remove this when push to `prod` version
	return value and value ~= ''
end

function export.empty(value)
	return value == nil or value == ''
end

return export

File: projects/inflection/modules/dev/dev_py/tools.py
import colorama
from colorama import Fore, Style

from projects.inflection.modules.dev.dev_py import mw
from projects.inflection.modules.dev.dev_py import utils as u


def trim_stress(str):
    # remove space after stress (we put it in the code just to make visually readable sources)
    return mw.ustring.gsub(str, '́ ', '́')


stash = {}
level = 0
colorama.init()


def clear_stash():
    global stash
    stash = {}


def add_stash(name, value):
    stash[name] = value


def apply_stash(str):
    for name, value in stash.items():
        str = mw.ustring.gsub(str, u.escape(name), value)

    return str


def replaced(str, pattern, replace_to):
    pattern = apply_stash(pattern)
    return mw.ustring.gsub(str, trim_stress(pattern), trim_stress(replace_to))


def replace(dict, key, pattern, replace_to):
    if key == 'all-sg':
        keys = ['gen-sg', 'dat-sg', 'prp-sg']  # without 'nom-sg', 'acc-sg' and 'ins-sg'
        for key in keys:
            dict[key] = replaced(dict[key], pattern, replace_to)

    elif key == 'all-pl':
        keys = ['nom-pl', 'gen-pl', 'dat-pl', 'ins-pl', 'prp-pl']  # without 'acc-pl'
        for key in keys:
            dict[key] = replaced(dict[key], pattern, replace_to)

    else:
        dict[key] = replaced(dict[key], pattern, replace_to)


def extract(str, pattern):
    pattern = apply_stash(pattern)
    return mw.ustring.match(str, trim_stress(pattern))


def check(string, values, checker):
    if type(values) == str:
        values = [values]

    for value in values:
        value = trim_stress(apply_stash(value))
        ok = checker == 'equals' and value == string \
             or checker == 'startswith' and mw.ustring.match(string, "^" + value) != None \
             or checker == 'endswith' and mw.ustring.match(string, value + "$") != None \
             or checker == 'penultimate' and mw.ustring.match(string, value + ".$") != None \
             or checker == 'contains' and mw.ustring.match(string, value) != None \
             or checker == 'contains_once' and len(mw.text.split(string, value)) == 2 \
             or checker == 'contains_several' and len(mw.text.split(string, value)) > 2
        if ok:
            return True

    return False


def equals(str, values):
    return check(str, values, 'equals')


def In(str, values):
    return equals(str, values)


def startswith(str, values):
    return check(str, values, 'startswith')


def endswith(str, values):
    return check(str, values, 'endswith')


def penultimate(str, values):
    return check(str, values, 'penultimate')


def contains(str, values):
    return check(str, values, 'contains')


def contains_once(str, values):
    return check(str, values, 'contains_once')


def contains_several(str, values):
    return check(str, values, 'contains_several')


def log(value):
    global level
    prefix = ' ' * level
    mw.log(prefix + value)


def green(value):
    return f'{Fore.GREEN}{value}{Style.RESET_ALL}'


def blue(value):
    return f'{Fore.BLUE}{value}{Style.RESET_ALL}'


def red(value):
    return f'{Fore.RED}{value}{Style.RESET_ALL}'


def green_line(func):
    def wrapped(*args, **kwargs):
        print(Fore.GREEN, end='')
        res = func(*args, **kwargs)
        print(Style.RESET_ALL, end='')
        return res
    return wrapped


def blue_line(func):
    def wrapped(*args, **kwargs):
        print(Fore.BLUE, end='')
        res = func(*args, **kwargs)
        print(Style.RESET_ALL, end='')
        return res
    return wrapped


def red_line(func):
    def wrapped(*args, **kwargs):
        print(Fore.RED, end='')
        res = func(*args, **kwargs)
        print(Style.RESET_ALL, end='')
        return res
    return wrapped


@green_line
def log_info(info):
    log('# ' + info)


# @blue_line
def call(module, name):
    log('@ ' + module + '.' + blue(name) + '()')
    global level


# @blue_line
def starts(module, name):
    log('↘ @ ' + module + '.' + blue(name) + '():')
    global level
    level += 4


def ends(module, name):
    global level
    level -= 4
    # log('↙ . ' + module + '.' + name + '()')
    log('↙ . ')


def returns(module, name, result):
    global level
    level -= 4
    # log('↙ . ' + module + '.' + name + '()')
    log('↙ . ')
    return result


# @red_line
def log_value(value, name):
    log('= ' + name + ': ' + red('"' + str(value) + '"'))


def log_table(t, name):
    log('- ' + name + ':')
    for key, value in t.items():
        log('  ["' + str(key) + '"] = ' + red('"' + str(value) + '"'))


def has_key(value, key=None):
    if key:
        return key in value
    return value


def has_value(value, key=None):
    if key:
        return key in value
    return value


def empty(value):
    return not value


# todo: log with colors (only for python)
# todo: log to files also (only for python)
# todo: create special short commands (like `u` and `pl`)
#  and use them from command line instead of running them via PyCharm
# todo: locally on python run all tests from lua tests data
# todo: try to get more tetsts from udarenieru html dumps... or from some zaliznyak file `All_Forms.txt`?
